src/audio/mixer.c

```diff
@@ -65,69 +65,164 @@ static void mix_n_s16(struct comp_dev *dev, struct audio_stream *sink,
 		      const struct audio_stream **sources, uint32_t num_sources,
 		      uint32_t frames)
 {
-	int16_t *src;
+	int16_t *src[PLATFORM_MAX_CHANNELS];
 	int16_t *dest;
 	int32_t val;
+	int nmax;
+	int n1;
+	int n2;
 	int i;
 	int j;
-	int channel;
-	uint32_t frag = 0;
-
-	for (i = 0; i < frames; i++) {
-		for (channel = 0; channel < sink->channels; channel++) {
-			val = 0;
-
-			for (j = 0; j < num_sources; j++) {
-				src = audio_stream_read_frag_s16(sources[j],
-								 frag);
-				val += *src;
+	int k;
+	int n;
+	int processed = 0;
+	int nch = sink->channels;
+	int samples = frames * nch;
+
+	dest = sink->w_ptr;
+	for (j = 0; j < num_sources; j++)
+		src[j] = sources[j]->r_ptr;
+
+	while (processed < samples) {
+		nmax = samples - processed;
+		n1 = audio_stream_bytes_without_wrap(sources[0], src[0]) >> 1; /* divide 2 */
+		n2 = audio_stream_bytes_without_wrap(sink, dest) >> 1;
+		for (i = 1; i < num_sources; i++) {
+			n = audio_stream_bytes_without_wrap(sources[i], src[i]) >> 1;
+			n1 = MIN(n, n1);
+		}
+		n = MIN(n1, n2);
+		n = MIN(n, nmax);
+		for (i = 0; i < n; i += nch) {
+			for (j = 0; j < nch; j++) {
+				val = *src[0];
+				src[0]++;
+				for (k = 1; k < num_sources; k++) {
+					val += *src[k];
+					src[k]++;
+				}
+				/* Saturate to 16 bits */
+				*dest = sat_int16(val);
+				dest++;
 			}
-
-			dest = audio_stream_write_frag_s16(sink, frag);
-
-			/* Saturate to 16 bits */
-			*dest = sat_int16(val);
-
-			frag++;
 		}
+		processed += n;
+		dest = audio_stream_wrap(sink, dest);
+		for (i = 0; i < num_sources; i++)
+			src[i] = audio_stream_wrap(sources[i], src[i]);
 	}
 }
 #endif /* CONFIG_FORMAT_S16LE */
 
-#if CONFIG_FORMAT_S24LE || CONFIG_FORMAT_S32LE
+#if CONFIG_FORMAT_S24LE
+/* Mix n 24 bit PCM source streams to one sink stream */
+static void mix_n_s24(struct comp_dev *dev, struct audio_stream *sink,
+		      const struct audio_stream **sources, uint32_t num_sources,
+		      uint32_t frames)
+{
+	int32_t *src[PLATFORM_MAX_CHANNELS];
+	int32_t *dest;
+	int64_t val;
+	int nmax;
+	int n1;
+	int n2;
+	int i;
+	int j;
+	int k;
+	int n;
```

Comment:
I know that we have a (supposedly soft) preference for 1 variable definition per line, but at least such short and supposedly homogeneous variables should be fine in fewer lines... Maybe at least you could put `i, j, k, n` and `n1, n2` in common lines. Not important of course

Reply:
OK, I think the guideline has been to declare one per line to not hide spilling the stack.